File: main/src/main/java/com/sedmelluq/discord/lavaplayer/tools/io/ThreadLocalHttpInterfaceManager.java
package com.sedmelluq.discord.lavaplayer.tools.io;

import org.apache.http.client.protocol.HttpClientContext;
import org.apache.http.impl.client.HttpClientBuilder;

/**
 * HTTP interface manager which reuses an HttpContext by keeping it as a thread local. In case a new interface is
 * requested before the previous one has been closed, it creates a new context for the returned interface. The HTTP
 * client instance used is created lazily.
 */
public class ThreadLocalHttpInterfaceManager extends AbstractHttpInterfaceManager {
  private final ThreadLocal<HttpInterface> httpInterfaces;

  /**
   * @param clientBuilder HTTP client builder to use for creating the client instance.
   */
  public ThreadLocalHttpInterfaceManager(HttpClientBuilder clientBuilder) {
    super(clientBuilder);

    this.httpInterfaces = ThreadLocal.withInitial(() ->
        new HttpInterface(getSharedClient(), HttpClientContext.create(), false)
    );
  }

  @Override
  public HttpInterface getInterface() {
    HttpInterface httpInterface = httpInterfaces.get();
    if (httpInterface.acquire()) {
      return httpInterface;
    }

    httpInterface = new HttpInterface(getSharedClient(), HttpClientContext.create(), false);
    httpInterface.acquire();
    return httpInterface;
  }
}

File: main/src/main/java/lavaplayer/tools/io/ThreadLocalHttpInterfaceManager.kt
package lavaplayer.tools.io

import lavaplayer.tools.http.HttpContextFilter
import lavaplayer.tools.http.SettableHttpRequestFilter
import org.apache.http.client.config.RequestConfig
import org.apache.http.client.protocol.HttpClientContext
import org.apache.http.impl.client.HttpClientBuilder

/**
 * HTTP interface manager which reuses an HttpContext by keeping it as a thread local. In case a new interface is
 * requested before the previous one has been closed, it creates a new context for the returned interface. The HTTP
 * client instance used is created lazily.
 *
 * @param clientBuilder HTTP client builder to use for creating the client instance.
 * @param requestConfig Request config used by the client builder
 */
class ThreadLocalHttpInterfaceManager(
    clientBuilder: HttpClientBuilder,
    requestConfig: RequestConfig
) : AbstractHttpInterfaceManager(clientBuilder, requestConfig) {
    private val httpInterfaces: ThreadLocal<HttpInterface>
    private val filter = SettableHttpRequestFilter()

    init {
        httpInterfaces = ThreadLocal.withInitial { HttpInterface(getSharedClient(), HttpClientContext.create(), false, filter) }
    }

    override fun get(): HttpInterface {
        val client = getSharedClient()

        var httpInterface = httpInterfaces.get()
        if (httpInterface.httpClient !== client) {
            httpInterfaces.remove()
            httpInterface = httpInterfaces.get()
        }

        if (httpInterface.acquire()) {
            return httpInterface
        }

        httpInterface = HttpInterface(client, HttpClientContext.create(), false, filter)
        httpInterface.acquire()
        return httpInterface
    }

    override fun setHttpContextFilter(filter: HttpContextFilter) {
        this.filter.set(filter)
    }
}
